WDL/runtime/cache.py

```diff
@@ -318,3 +330,43 @@ def clean(line: int, column: int = 1, end_column: Optional[int] = None) -> List[
             clean(pos.end_line, 1, pos.end_column),
         )
     )
+
+
+class FileCoherence(abc.ABC):
+    _logger: logging.Logger
+
+    def __init__(self, logger):
+        self._logger = logger.getChild("FileCoherence")
+        self.cache_file_modification_time = 0.0
+
+    def check_output_files(self, cache_file_path: str, cache_output: Env.Bindings[Base]) -> bool:
+        if self.cache_file_modification_time == 0.0:
+            self.cache_file_modification_time = self.get_last_modified_time(cache_file_path)
+        for output in cache_output:
+            if type(output._value) == File:
```

Comment:
We need to handle if the values include Files inside compound types, like an `Array[File]` or other more-complex cases. The helper function [`WDL.Value.rewrite_env_files()`](https://miniwdl.readthedocs.io/en/latest/WDL.html#WDL.Value.rewrite_env_files) can help with the recursive traversal.